include/tvm/runtime/profiling.h

```diff
@@ -0,0 +1,156 @@
+   * them before calling `SyncAndGetElapsedNanos` on any of them.
+   *
+   * Note: this function should be only called once per object. It may incur
+   * a large synchronization overhead (for example, with GPUs).
+   */
+  virtual int64_t SyncAndGetElapsedNanos() = 0;
+
+  virtual ~TimerNode() {}
+
+  static constexpr const char* _type_key = "TimerNode";
+  TVM_DECLARE_BASE_OBJECT_INFO(TimerNode, Object);
+};
+
+/*! \brief Timer for a specific device.
+ *
+ * You should not construct this class directly. Instead use `StartTimer`.
+ */
+class Timer : public ObjectRef {
+ public:
+  /*! \brief Start the timer.
+   *
+   * Note: this function should only be called once per object.
+   */
+  void Start() { operator->()->Start(); }
+  /*! \brief Stop the timer.
+   *
+   * Note: this function should only be called once per object.
+   */
+  void Stop() { operator->()->Stop(); }
+  /*! \brief Synchronize timer state and return elapsed time between `Start` and `Stop`.
+   * \return The time in nanoseconds between `Start` and `Stop`.
+   *
+   * This function is necessary because we want to avoid timing the overhead of
+   * doing timing. When using multiple timers, it is recommended to stop all of
+   * them before calling `SyncAndGetElapsedNanos` on any of them.
+   *
+   * Note: this function should be only called once per object. It may incur
+   * a large synchronization overhead (for example, with GPUs).
+   */
+  int64_t SyncAndGetElapsedNanos() { return operator->()->SyncAndGetElapsedNanos(); }
+  TVM_DEFINE_MUTABLE_OBJECT_REF_METHODS(Timer, ObjectRef, TimerNode);
+};
+
+/*!
+ * \brief Default timer if one does not exist for the context.
+ * \param ctx The context to time on.
+ *
+ * Note that this timer performs synchronization between the device and CPU,
+ * which can lead to overhead in the reported results.
+ */
+Timer DefaultTimer(TVMContext ctx);
+
+/*!
+ * \brief Get a device specific timer.
+ * \param ctx The device context to time.
+ * \return A `Timer` that has already been started.
+ *
+ * Use this function to time runtime of arbitrary regions of code on a specific
+ * device. The code that you want to time should be running on the device
+ * otherwise the timer will not return correct results. This is a lower level
+ * interface than TimeEvaluator and only runs the timed code once
+ * (TimeEvaluator runs the code multiple times).
+ *
+ * Example usage:
+ * \code{.cpp}
+ * Timer t = StartTimer(TVMContext::cpu());
+ * my_long_running_function();
+ * t.Stop();
+ * ... // some more computation
+ * int64_t nanosecs = t.SyncAndGetElapsedNanos() // elapsed time in nanoseconds
+ * \endcode
+ *
+ * To add a new device-specific timer, register a new function
+ * "profiler.timer.my_device" (where `my_device` is the `DeviceName` of your
+ * device). This function should accept a `TVMContext` and return a new `Timer`
+ * that has already been started.
+ */
+inline Timer StartTimer(TVMContext ctx) {
```

Comment:
How about just make a constructor `explicit Timer(ctx);` ?

Reply:
Seems inline function is not necessary and can make as a TVM_DLL function

Reply:
Having a static member makes sense. Done.